heareval/tasks/pipeline.py

```diff
@@ -284,18 +317,49 @@ def get_metadata(self):
         )
         return metadata
 
+    def subsample_metadata(self, metadata, max_files):
+
+        # Get group count for each group
+        grp_count = metadata["stratify_key"].value_counts() * max_files / len(metadata)
+        # Sort by the split key and the subsample key
+        metadata = metadata.sort_values(
+            by=["split_key", "subsample_key"], ascending=[True, True]
+        )
+        # Groupby and select the required sample from each group
+        sampled_metadata = pd.concat(
+            [
+                # Ensure at least 1 sample is selected for each group
+                stratify_grp.head(1 if grp_count[grp] < 1 else int(grp_count[grp]))
+                for grp, stratify_grp in metadata.groupby("stratify_key")
+            ]
+        )
+        # Do some assertions
+        assert set(sampled_metadata["stratify_key"].unique()) == set(
+            metadata["stratify_key"].unique()
+        ), "All stratify groups are not in the sampled metadata."
+        # Add the num of groups here for handling addition in case of highly
+        # imbalanced subsampling
+        assert len(sampled_metadata) <= max_files + len(
+            grp_count
+        ), "Sampled metadata is more than the allowed max files + unique groups"
+
+        print(f"Datapoints in split after resampling: {len(sampled_metadata)}")
+        return sampled_metadata
+
     def run(self):
 
         metadata = self.get_metadata()
         num_files = len(metadata)
         max_files = num_files if self.max_files is None else self.max_files
         if num_files > max_files:
-            print(f"{num_files} audio files in corpus, keeping only {max_files}")
-
-        # Sort by the subsample key and select the max_files number of samples
-        metadata = metadata.sort_values(
-            by=["split_key", "subsample_key"], ascending=[True, True]
-        ).iloc[:max_files]
+            print(
+                f"{num_files} audio files in corpus."
+                "Max files to subsample: {max_files}"
+            )
+            metadata = self.subsample_metadata(metadata, max_files)
+            assert self.subsample_metadata(metadata, max_files).equals(
+                metadata
+            ), "The subsampling is not stable"
```

Comment:
Check
- Is the subsampling stable across two runs?

Reply:
Okay, should this also be a unit test?

Reply:
Yes. we can include it in the unit test for the pipeline. I have done some fixes and made it more robust for now here.